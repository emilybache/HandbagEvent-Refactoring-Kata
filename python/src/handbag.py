import dataclasses

# Max handbag size in litres - larger than this and it's counted as luggage
MAX_HANDBAG_CAPACITY = 25.0


@dataclasses.dataclass
class Event:
    type: int
    size: int
    status_code: int
    checksum: int
    payload: str
    has_checksum: bool


@dataclasses.dataclass
class Handbag:
    compartments: int
    designer: str
    market_segment: str
    strap_type: str
    capacity: float


def predict_sales(handbag: Handbag):
    score = 0
    if handbag.compartments > 10:
        score -= 1
    if handbag.designer in ["H&M", "Zara", "Esprit"]:
        score += 3
    if handbag.market_segment == "Trendy":
        score += 1
    if handbag.market_segment == "Traveller":
        score -= 2
    if handbag.strap_type == "Clutch":
        score -= 1
    if handbag.strap_type == "Shoulder":
        score += 1
    if handbag.capacity < 1:
        score -= 1
    if handbag.capacity > 3:
        score += 1
    return score


def fits_desired_market_segment(handbag: Handbag):
    predicted_sales = predict_sales(handbag)
    return predicted_sales > 0


def has_recognized_brand(handbag: Handbag):
    return handbag.designer in ["H&M", "Zara", "Esprit"]


def has_dead_brand(handbag: Handbag):
    return handbag.designer in ["Debenhams", "Woolworths"]


def init_handbag_event(handbag: Handbag, event: Event) -> bool:
    result = False
    size = 0
    if handbag.capacity <= MAX_HANDBAG_CAPACITY and event is not None:
        size = 1
        if fits_desired_market_segment(handbag):
            if has_recognized_brand(handbag) or has_dead_brand(handbag):
                response = determine_event_payload(handbag, event)
                if response:
                    size = len(response)
                    result = True
            else:
                event.status_code = "Invalid brand"
                result = True
        else:
            event.status_code = "Invalid market segment"
            result = True
    else:
        # timeout
        pass

    if result:
        event.size = size
        event.checksum = calculate_checksum(event)
        event.has_checksum = True
        result = True
    return result


def determine_event_payload(handbag: Handbag, event: Event):
    event.payload = None
    if handbag.strap_type == "Shoulder":
        event.payload = "Shoulder Handbag"
        event.status_code = "Good"
        event.type = 1
    elif handbag.strap_type == "Clutch":
        event.payload = "Clutch Handbag"
        event.status_code = "Good"
        event.type = 2
    elif handbag.strap_type == "Fixed":
        event.payload = "Hard strap Handbag"
        event.status_code = "Good"
        event.type = 3
    else:
        event.status_code = "Unknown"
    return event.payload


def calculate_checksum(event):
    return (event.type + len(event.status_code) + event.size + len(event.payload)) % 10
