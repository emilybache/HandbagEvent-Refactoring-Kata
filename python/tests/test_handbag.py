import approvaltests

from handbag import *


def print_event(success: bool, event: Event) -> str:
    if event is None:
        return f"Success: {success}, Event: None"
    return f"""
Success: {success}
Event(
    type: {event.type}
    size: {event.size}
    status_code: {event.status_code}
    checksum: {event.checksum}
    payload: {event.payload}
    has_checksum: {event.has_checksum}
)
    """


def test_trendy_shoulder_bag():
    input = Handbag(1, "H&M", "Trendy", "Shoulder", 0.55)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_trendy_clutch_bag():
    input = Handbag(1, "Zara", "Trendy", "Clutch", 0.2)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_trendy_fixed_bag():
    input = Handbag(1, "Esprit", "Trendy", "Fixed", 0.2)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_bad_market_fit():
    input = Handbag(1, "Debenhams", "Trendy", "Clutch", 0.35)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_dead_brand():
    input = Handbag(1, "Debenhams", "Trendy", "Shoulder", 4)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_luggage():
    input = Handbag(4, "Samsonite", "Traveller", "Trolley", 60)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_small_luggage():
    input = Handbag(2, "Samsonite", "Trendy", "Box", 10)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_many_compartments():
    input = Handbag(14, "Esprit", "Traveller", "Rucksack", 20)
    output = Event(0, 0, 0, 0, "", False)
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))


def test_missing_output():
    input = Handbag(1, "H&M", "Trendy", "Shoulder", 0.55)
    output = None
    success = init_handbag_event(input, output)
    approvaltests.verify(print_event(success, output))
