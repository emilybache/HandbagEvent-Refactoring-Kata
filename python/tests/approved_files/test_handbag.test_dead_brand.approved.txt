
Success: True
Event(
    type: 1
    size: 16
    status_code: Good
    checksum: 7
    payload: Shoulder Handbag
    has_checksum: True
)
    
